stetho/src/main/java/com/facebook/stetho/inspector/protocol/module/DOM.java

```diff
@@ -2,30 +2,315 @@
 
 package com.facebook.stetho.inspector.protocol.module;
 
+import android.util.SparseArray;
+
+import com.facebook.stetho.common.Util;
+import com.facebook.stetho.inspector.helper.ChromePeerManager;
 import com.facebook.stetho.inspector.jsonrpc.JsonRpcPeer;
 import com.facebook.stetho.inspector.jsonrpc.JsonRpcResult;
 import com.facebook.stetho.inspector.protocol.ChromeDevtoolsDomain;
 import com.facebook.stetho.inspector.protocol.ChromeDevtoolsMethod;
+import com.facebook.stetho.json.ObjectMapper;
 import com.facebook.stetho.json.annotation.JsonProperty;
 
 import org.json.JSONObject;
 
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.GuardedBy;
+
 public class DOM implements ChromeDevtoolsDomain {
-  public DOM() {
+  private final Provider mDOMProvider;
+  private final DOMPeerManager mPeerManager;
+  private final ObjectMapper mObjectMapper;
+
+  private final Object mNodeMapSync = new Object();
+
+  @GuardedBy("mNodeMapSync")
+  private int mNextNodeId = 1;
+
+  @GuardedBy("mNodeMapSync")
+  private final Map<Object, Integer> mElementToNodeIdMap = new HashMap<>();
+
+  @GuardedBy("mNodeMapSync")
+  private final SparseArray<Object> mNodeIdToElementMap = new SparseArray<>();
+
+  public DOM(Provider provider) {
+    mDOMProvider = Util.throwIfNull(provider);
+    mDOMProvider.setListener(new ProviderListener());
+
+    mPeerManager = new DOMPeerManager();
+    mObjectMapper = new ObjectMapper();
+  }
+
+  private static final class DOMPeerManager extends ChromePeerManager {
```

Comment:
Separate class.

Also, I'm thinking you should invoke `mDOMProvider.setListener` during `onFirstPeerRegistered`.  See `DatabasePeerManager` for reference.  This could be a good opportunity to upgrade the `setListener` API to `addListener`/`removeListener` as well.